dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/InMemoryQueryEngine.java

```diff
@@ -117,62 +112,77 @@ private <T extends IdentifiableObject> List<T> runSorter(Query query, List<T> ob
     return sorted;
   }
 
-  private <T> boolean matches(Query query, T object) {
+  private <T extends IdentifiableObject> boolean matches(
+      Query<T> query, T object, List<Matcher> matchers) {
     if (query.getRootJunctionType() == Junction.Type.OR) {
       // OR
-      for (Filter filter : query.getFilters()) {
-        if (matches(query, object, filter)) return true;
-      }
+      for (Matcher matcher : matchers) if (matches(query, object, matcher)) return true;
       return false;
     }
     // AND
-    for (Filter filter : query.getFilters()) {
-      if (!matches(query, object, filter)) return false;
-    }
+    for (Matcher matcher : matchers) if (!matches(query, object, matcher)) return false;
     return true;
   }
 
-  private <T> boolean matches(Query query, T object, Filter filter) {
+  /**
+   * The main purpose of matchers is to "cache" the knowledge of how to extract a value for a filter
+   * test from an object of the listed object type.
+   *
+   * @param filter the filter that is applied
+   * @param value a function to extract the filter property value from an object
+   */
+  private record Matcher(Filter filter, Function<IdentifiableObject, Object> value) {}
```

Comment:
FYI: ATM the `Matcher` indirection adds no benefit. I started this and then found that the refactoring the `getValue` needs will be quite extensive so I rather do it in another PR. But this is a preparation for it.